File: Rust/collatz.rs
///////////////////////////////////////////////////////////////////////
/////////          COLLATZ CONJECTURE PROJECT             /////////////
//                                                                    /  
// A program to find collatz sequence length of ten smallest integers /
// in a given range and prints them in descending order of their      /
// respective sequence lengths and the integers themselves            /
///////////////////////////////////////////////////////////////////////
use std::env;
use std::process::exit;
// A global struct
    #[derive(Copy, Clone)]
    struct Numpair {
       num: i64,     // An integer
       seqlen: i64   // Collatz sequence length of the integer
     }
 
fn main() {
    // Get arguments from command line
    let args: Vec<_> = env::args().collect();
    let mut a1: i64 = args[1].to_string().parse::<i64>().unwrap_or(0);
    let mut a2: i64 = args[2].to_string().parse::<i64>().unwrap_or(0);  
    // Initialize an array of structs (size 10) with all structs with their fields set to zero 
    let mut numpairs: [Numpair; 10] = [Numpair{num: 0, seqlen: 0}; 10];
    
    check_args(&mut a1, &mut a2);
    find_sequence_lengths_in_range(a1, a2, &mut numpairs);
}  

/*
  Verifies the arguments and quits if they are invalid
  @param n1 - first integer argument
  @param n2 - second integer argument
*/
fn check_args(n1: &mut i64, n2: &mut i64) {
   // Reject same arguments and those less than 1
   if *n1 == *n2 ||  *n1 < 1 || *n2 < 1 {
      println!("Invalid range and/or arguments!\nExiting...");
      exit(1);
   }
   // Swap arguments if needed.
   if *n1 > *n2 {
      let tmp = *n1;
      *n1 = *n2;
      *n2 = tmp;
   }
}

/*
  Finds the length of the collatz sequence
  @param num - an integer
  @return steps - the length of sequence
*/
fn find_sequence_length(num :i64) -> i64 {
   let (mut steps, mut m): (i64, i64) = (0, num);
   
   while m != 1 {
      steps += 1;
      if m%2 == 0 {
         m = m / 2;
      } else {
         m = 3*m + 1;
      }
   }
   return steps; 
}

/*
  Finds sequence lengths of numbers in a given range (inclusive)
  @param n1 - integer at start of range
  @param n2 - integer at end of range
  @paran arr - a mutable reference to our array 
  @note This reference to the array will be passed onto the another
  function but it needed to be passed here first to remain in scope
*/
fn find_sequence_lengths_in_range(n1: i64, n2: i64, arr: &mut[Numpair]) {
   let mut apair: Numpair = Numpair{num: 0, seqlen: 0};
   let mut i = 0; // For keeping count of iterations
     
   for n in n1..n2+1 {
      apair.num = n; 
      apair.seqlen = find_sequence_length(n);
      i += 1;
      add_to_array(arr, apair, &mut i);       
   }
   
   println!("Sorted based on sequence length");
   arr.sort_by_key(|x| x.seqlen);
   arr.reverse();
   print_numbers_with_length(arr);
   println!("Sorted based on integer size"); 
   arr.sort_by_key(|x| x.num);
   arr.reverse();
   print_numbers_with_length(arr);    
}

/*
  Adds a Numpair struct into an array based on some specific conditions.
  @param arr - the array where an element is to added (or not)
  @param elem - element to be added in the array
  @param i - index of array 
*/
fn add_to_array(arr: &mut[Numpair], elem: Numpair, i: &mut i64) {
   let mut same_len = false;
   if *i == 1 { // Adding first element
      arr[0] = elem;
   } else if *i <= 10 { // Adding to partially filled array
      for x in 0..*i {  // Recall that the ending value is exclusive
         if elem.seqlen == arr[x as usize].seqlen {
            same_len = true;
            *i -= 1; // Since current index will remain empty
            break; 
         }        
      }
      if !same_len {
         arr[*i as usize - 1] = elem;
      }
      if *i == 10 {
         arr.sort_by_key(|x| x.seqlen);
         arr.reverse();
      } 
   } else { // When array is full
      for x in 0..10 {  // Recall that the ending value is exclusive
         if elem.seqlen == arr[x as usize].seqlen {
            same_len = true;
            break;
         }
      }
      if !same_len && elem.seqlen > arr[9].seqlen {
         arr[9] = elem; 
         arr.sort_by_key(|x| x.seqlen);
         arr.reverse();
      }
   }    
}

/*
  Print the array elements according to the requirements
  @param arr - the array to be printed
*/
fn print_numbers_with_length(arr: &mut[Numpair]) {
   for i in 0..10 {
      if arr[i as usize].num == 0 {
         break;
      }
      println!("{0:>21} {1:>20}", arr[i as usize].num, arr[i as usize].seqlen);
   }
} 

File: Julia/collatz.jl
#!/usr2/local/julia-1.8.2/bin/julia
#######################################################################
#  A program to find collatz sequence length of ten smallest integers #
#  in a given range and prints them in descending order of their      #
#  respective sequence lengths and the integers themselves            #
#######################################################################

# A global struct
struct Numpair
   num::Int64     # An integer
   seqlen::Int64  # Collatz sequence length of the integer
end
   
"""
Verifies the arguments and quits if they are invalid.
Passes them on to the next function if they are valid.
# Arguments
- 'n1::Integer`: first argument
- `n2::Integer`:  second argument
"""
function check_args_and_continue(n1, n2)
   # Reject same arguments and those less than 1
   if n1 == n2 ||  n1 < 1 ||  n2 < 1 
      println("Invalid range and/or arguments!\nExiting...")
      return
   end
   # Swap arguments if needed before passing them on
   if n1 > n2 
      tmp = n1
      n1 = n2
      n2 = tmp
   end
   find_sequence_lengths_in_range(n1, n2)
end

"""
Finds the length of the collatz sequence
# Arguments
- `num::Integer`: the number for which the collatz sequence length
                  is calculated
# Returns 
- `steps::Integer`: the length of sequence
"""
function find_sequence_length(num)
   m = num
   steps::Int64 = 0
   while m != 1
      steps += 1
      if m%2 == 0
         m = m / 2
      else
         m = 3*m + 1
      end
   end
   return steps
end       
   
"""
Finds sequence lengths of numbers in a given range (inclusive)
# Arguments
- `n1::Integer`: start of range
- `n2::Integer`: end of range
"""
function find_sequence_lengths_in_range(n1, n2)  
   numpairs = Array{Numpair}(undef, 0) # Declare an empty vector of Numpair structs       
   for num = n1:n2 
      apair = Numpair(num, find_sequence_length(num))
      add_to_array(numpairs, apair)       
   end 
   println("Sorted based on sequence length") 
   sort!(numpairs, by = x -> x.seqlen, rev=true)     
   print_numbers_with_lengths(numpairs)
   println("Sorted based on integer size")
   sort!(numpairs, by = x -> x.num, rev=true)
   print_numbers_with_lengths(numpairs)
   return nothing
end
         
"""
Adds a Numpair struct into an array based on some specific conditions.
# Arguments
- `arr::Array{Numpair}`: the array where an element is to added (or not)
- `elem::Numpair`:  element to be added in the array
# Note
  Vector in Julia are just 1-D arrays, arrays in Julia are always dynamic.
"""
function add_to_array(arr, elem)
   samelen = false
   # Adding first element
   if length(arr) == 0
      push!(arr, elem)
      # Adding until array has 10 elements
   elseif length(arr) < 10 
      for x in arr 
         if elem.seqlen == x.seqlen
            samelen = true
            break
         end 
      end
      if !samelen 
         push!(arr, elem)
      end 
      if length(arr) == 10
         sort!(arr, by = x -> x.seqlen, rev=true)
      end
   # When array is filled with 10 elements
   else 
      for x in arr
         if elem.seqlen == x.seqlen 
            samelen = true
            break
         end
      end
      if !samelen && elem.seqlen > arr[10].seqlen
         arr[10] = elem
         sort!(arr, by = x -> x.seqlen, rev=true)
      end        
   end
   return nothing
end
   
"""
Print the array elements according to the requirements
# Arguments
- `arr::Array{Numpair}`: the array to be printed
"""
function print_numbers_with_lengths(arr)
   for x in arr
      println(lpad(x.num,21," "), lpad(x.seqlen,21," "))
   end
   return nothing
end
   
### Main starts here

try   
   # Read command line arguments
   a1 = parse(Int64, ARGS[1])
   a2 = parse(Int64, ARGS[2])
   check_args_and_continue(a1, a2)
catch e
   println("Invalid range and/or arguments!\nExiting...")
end

exit()
